Add launcher for JITServer

ashu-mehra opened the pull request:
Added a launcher for JITServer which starts the JVM with `-XX:StartAsJITServer` option.
It also adds some GC options which make sense for JITServer.

Fixes https://github.com/eclipse/openj9/issues/7800

Signed-off-by: Ashutosh Mehra <mehra.ashutosh@ibm.com>

ashu-mehra commented:
@mpirvu @DanHeidinga this PR not yet complete as I need to add the make files to build it, but the launcher code is ready.

ashu-mehra commented:
@DanHeidinga I am assuming the launcher would be placed under `jre/bin` (for java 8) or `bin` (java 9+). Is that correct?

ashu-mehra commented:
Added `module.xml` and `CMakeLists.txt` files. Currently the executable `jitserver` is placed alongside the VM libraries under `vm` directory when doing a build.

mpirvu commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,308 @@
+		rc = append(jvmLibPath, J9_MAX_PATH, "/lib/j9vm/libjvm.so");
+	}
+	if (0 != rc) {
+		fprintf(stderr, "Path exceeded maximum length\n");
```
It would be useful to the user to know the maximum value for the path length. Let's add it to this message

mpirvu commented on runtime/jitserver_launcher/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+################################################################################
+
+add_executable(jitlauncher
+	jitlauncher.c
```
Is this supposed to be `jitserver.c`?

DanHeidinga commented:
> I am assuming the launcher would be placed under `jre/bin` (for java 8) or `bin` (java 9+). Is that correct?

For JDK 8, I'd expect it in both `bin/` and `jre/bin` like other executables.

For JDK11+, `bin` is the right location

DanHeidinga commented:
See also my comment in https://github.com/eclipse/openj9/issues/7800#issuecomment-561307743 for how this should evolve

DanHeidinga commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,358 @@
+	if (-1 == rc) {
+		fprintf(stderr, "readlink(%s) failed, error code=%d\n", SELF_EXE, errno);
+		goto _end;
+	}
```
```suggestion
	} else if (rc > (bufLen - 1)) {
		/* Link can be changed between reading it and the stat call */
		fprintf("readlink(%s) failed as link was changed\n", SELF_EXE");
		free(buffer);
		buffer = NULL;
		goto _end;
	}
```

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,308 @@
+}
+
+static int32_t
+isJava8(char *jdkOrJreHome)
```
is there a macro available for that?

keithc-ca replied:
Yes: `JAVA_SPEC_VERSION == 8`.

ashu-mehra commented:
@mpirvu I had to rollback my changes for making the main thread as the listener thread. I realised it won't be a good idea as the main thread would then be blocked forever. I have removed that code, and made the main thread join the listener thread to prevent it from exiting.
Unfortunately the VM API `createThreadWithCategory` does not create a join-able thread. Not sure why is that so. I have to add another API `createJoinableThreadWithCategory` which is used for creating the listener thread. Then the main thread can do a join on it.

@DanHeidinga  hopefully the JITServer interface and jitlauncher is now in line with the suggestions made in #7827.

DanHeidinga commented on runtime/oti/j9arch.h:
```diff
@@ -0,0 +1,53 @@
+#endif
+
+#define OPENJ9_CR_JVM "compressedrefs"
+#define OPENJ9_NOCR_JVM "default"
```
Can you add the _DIR into these names again?

DanHeidinga commented on runtime/oti/jitserver_api.h:
```diff
@@ -0,0 +1,47 @@
+typedef struct JITServer {
+	jint (JNICALL *StartJITServer)(struct JITServer *);
+	jint (JNICALL *WaitForTermination)(struct JITServer *);
+	JavaVM *jvm;
```
I would prefer to see these function ptrs named with lower case names

DanHeidinga commented on runtime/vm/threadhelp.cpp:
```diff
@@ -298,6 +299,66 @@ getMonitorForWait(J9VMThread* vmThread, j9object_t object)
+
+        if (J9THREAD_SUCCESS != omrthread_attr_init(&attr)) {
+                return J9THREAD_ERR_CANT_ALLOC_CREATE_ATTR;
+        }
```
tabs vs spaces

DanHeidinga commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,402 @@
+	int32_t i = 0;
+	int32_t argIndex = 0;
+	int32_t numJvmOptions = 0;
+	//jniCreateJavaVM createJavaVM = NULL;
```
```suggestion
```

ashu-mehra replied:
Done

mpirvu commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1518,77 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+	JNIEnv *env = NULL;
+	int32_t rc = 0;
+
+	rc = (*vm)->AttachCurrentThread(vm, (void **)&env, NULL);
```
I am assuming that the attach operation is needed at some level. Is there a reason to detach immediately after `startJITServer`? I am asking because in the next operation (`WaitForTermination`) we again attach and then detach.

ashu-mehra replied:
It is not guaranteed that the user of `JITServer` would call `startJITServer` and `waitForTermination` on same thread. And from user perspective `JITServer` is not a JVM, but since we are internally creating a JVM we should hide the details that the creation and usage of JVM requires, like these attach and detach calls. Does that make sense?

mpirvu replied:
I am ok with the implementation in the JIT related files

ashu-mehra replied:
Looks like some how I managed to delete my answer to the first comment. For the record adding it again:
```It is possible that the calls to `startJITServer` and `WaitForTermination` are on different threads. Also the JITServer is not same as the JVM, but since we are internally creating a JVM, we should hide the details involved in creating and using JVM like these attach an detach calls.```

ashu-mehra commented on runtime/j9vm/j9vmnatives.xml:
```diff
@@ -25,6 +25,9 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<export name="JNI_GetDefaultJavaVMInitArgs" />
+		<export name="JITServer_CreateServer" />
+		<export name="StartJITServer" />
+		<export name="WaitForTermination" />
```
Done

ashu-mehra commented on runtime/j9vm/exports.cmake:
```diff
@@ -40,6 +40,9 @@ jvm_add_exports(jvm
 	JNI_GetDefaultJavaVMInitArgs
+	JITServer_CreateServer
+	StartJITServer
+	WaitForTermination
```
Done

keithc-ca replied:
`StartJITServer` and `WaitForTermination` are still being exported here: why? The autotools variant does not.

ashu-mehra commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1518,77 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+		rc = JIT_SERVER_OOM;
+		goto _end;
+	}
+	*jitServer = server;
```
Done

ashu-mehra commented:
@DanHeidinga I have updated the commit with suggested changes. Please review.

ashu-mehra commented:
I forgot to add that `omrthread_attr_set_detachstate` is not exported by omr. I have opened a PR https://github.com/eclipse/omr/pull/4658 to fix that. Until that OMR PR is merged, I will keep it as WIP.

ashu-mehra commented:
eclipse/omr#4658 is merged. Removed the WIP tag.

mpirvu commented:
JITServer tests are no longer possible with this PR. I'll attempt a compilation test.

mpirvu commented:
Jenkins compile xlinuxjit jdk8

mpirvu commented:
```

mpirvu commented:
I thought the omr change has propagated:
```
13:38:04  ../..//libj9vm29.so: undefined reference to `omrthread_attr_set_detachstate'
13:38:04  collect2: error: ld returned 1 exit status
13:38:04  ../../makelib/targets.mk:382: recipe for target '../../vmtest' failed
13:38:04  gmake[3]: *** [../../vmtest] Error 1
13:38:04  makefile:244: recipe for target 'vmtest' failed
13:38:04  gmake[2]: *** [vmtest] Error 2
13:38:07  /home/jenkins/workspace/Build_JDK8_x86-64_linux_jit_Personal/closed/OpenJ9.gmk:509: recipe for target 'build-j9' failed
13:38:07  gmake[1]: *** [build-j9] Error 2
13:38:07  /home/jenkins/workspace/Build_JDK8_x86-64_linux_jit_Personal/closed/make/Main.gmk:39: recipe for target 'j9vm-build' failed
13:38:07  make: *** [j9vm-build] Error 2
```

ashu-mehra commented:
@mpirvu It has not yet propagated into `openj9` branch of `openj9-omr`.

keithc-ca commented on runtime/compiler/runtime/Listener.cpp:
```diff
@@ -82,7 +84,8 @@ static int32_t J9THREAD_PROC listenerThreadProc(void * entryarg)
 void TR_Listener::startListenerThread(J9JavaVM *javaVM)
    {
-   PORT_ACCESS_FROM_JAVAVM(javaVM); 
+#if 0
```
I suggest `runListenLoop` is better `start` is what you do to a thread (which executes in parallel with the caller).

ashu-mehra replied:
@keithc-ca this code is outdated now. I have changed back to original code that does start a new thread.

keithc-ca commented on runtime/j9vm/j9vmnatives.xml:
```diff
@@ -25,6 +25,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<export name="JNI_CreateJavaVM" />
 		<export name="JNI_GetCreatedJavaVMs" />
 		<export name="JNI_GetDefaultJavaVMInitArgs" />
+		<export name="JITServer_CreateServer" />
```
This should not be necessary (see comment above).

ashu-mehra replied:
I think we do need to export `JITServer_CreateServer`, the others are not required as they are accessed as function pointers. `JITServer_CreateServer` can be called directly by the user.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+#endif
+
+	/* Consume any trailing '/'. We should restore it before exit. */
+	if (path[pathLen - 1] == DIR_SEPARATOR_CHAR) {
```
This should not assume `pathLen > 0`.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+static int32_t 
+append(char *buffer, int32_t bufLen, char *str)
+{
+	if (strlen(buffer) + strlen(str) + 1 > bufLen) {
```
`+ 1 >` can be simplified to `>=`.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+        if (0 != rc) {
+                fprintf(stderr, "Path exceeded maximum allowed length of %d\n", OPENJ9_MAX_PATH);
+                goto _end;
+        }
```
Inconsistent indentation.

keithc-ca commented on runtime/oti/j9arch.h:
```diff
@@ -0,0 +1,53 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
```
This file needs `#ifdef j9arch_h` re-include protection.

keithc-ca replied:
It should also have `extern "C"` for C++ users.

keithc-ca commented on runtime/vm/threadhelp.cpp:
```diff
@@ -298,39 +299,84 @@ getMonitorForWait(J9VMThread* vmThread, j9object_t object)
+	if (J9THREAD_SUCCESS != rc) {
 		goto destroy_attr;
-	}
+	};
```
Why the added semicolon?

keithc-ca commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1520,80 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+}
+
+int32_t JNICALL JITServer_CreateServer(JITServer **jitServer, void *vm_args) {
+	JNIEnv *env = NULL;
```
Coding style throughout: function return type and opening `{` should each be on separate lines:
```
jint JNICALL
JNI_CreateJavaVM(JavaVM **pvm, void **penv, void *vm_args)
{
```

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+#include "jitserver_api.h"
+#include "jitserver_error.h"
+
+#define DIR_SEPARATOR_CHAR '/'
```
This whole file assumes Linux: I think an explicit check of that is appropriate and an `#error` otherwise.

pshipton replied:
It doesn't seem Keith's comment has been addressed.

ashu-mehra replied:
aah! I missed this one. I will add the check.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+	int32_t rc = 0;
+
+ 	/* pathBuffer can be either <java_home> or <java_home>/jre */
+	rc = checkLastLevel(pathBuffer, "jre");
```
This test is only appropriate for `JAVA_SPEC_VERSION == 8`.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,400 @@
+		goto _end;
+	}
+	strcpy(jvmLibPath, selfPath);
+	free(selfPath);
```
Only one of `jvmLibPath` or `selfPath` is required.

Also, this computation and the collection of support functions is needlessly complicated.

ashu-mehra commented:
Is there a way to ignore `jitserver_launcher` in UMA based build system if `JITSERVER_SUPPORT` env variable is not set. With cmake, I can do something like this:
```
+if (JITSERVER_SUPPORT)
+       add_subdirectory(jitserver_launcher)
+endif()
```
Is there a way to achieve the same in uma?

keithc-ca commented:
> Is there a way to ignore `jitserver_launcher` in UMA based build system if `JITSERVER_SUPPORT` env variable is not set. With cmake, I can do something like this:
> 
> ```
> +if (JITSERVER_SUPPORT)
> +       add_subdirectory(jitserver_launcher)
> +endif()
> ```
> 
> Is there a way to achieve the same in uma?

I don't think environment variables are a good way to communication configuration choices.

The way that has been done so far is via a flag in the build spec; it sounds like you want to add and use `module_jitserver` or similar. See `buildspecs/j9.flags`.

ashu-mehra commented on runtime/jitserver_launcher/module.xml:
```diff
@@ -0,0 +1,49 @@
+		<include-if condition="spec.linux_390.*"/>
+		<include-if condition="spec.linux_ppc.*"/>
+		<include-if condition="spec.linux_x86.*"/>
+		<include-if condition="spec.flags.build_jitserver"/>
```
This doesn't seem to work. Even when `build_jitserver` is not set in the spec, `jitserver` executable gets generated. @DanHeidinga @keithc-ca any idea why is that so?

keithc-ca replied:
The `spec.linux` conditions should be redundant since you added the `build_jitserver` flag.

Are you adjusting the `build_jitserver` flag based on the configuration option, like [1]?

[1] https://github.com/ashu-mehra/openj9-openjdk-jdk11/blob/openj9/closed/OpenJ9.gmk#L357-L364

ashu-mehra replied:
> Are you adjusting the build_jitserver flag based on the configuration option,

Yes, created a PR for `OpenJ9.gmk` changes https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/350

keithc-ca replied:
When I build with this and jdk8/350 and run `configure` without `--enable-jitserver`, no makefile is generated in `vm/jitserver_launcher`, signalling the desired behavior. After saying `make images`, there's no file called `jitserver` anywhere.

ashu-mehra replied:
Not sure what went wrong in my case. Anyway good if it is working. Thanks for trying it out.

keithc-ca commented on runtime/j9vm/exports.cmake:
```diff
@@ -350,3 +350,7 @@ if(NOT JAVA_SPEC_VERSION LESS 14)
 endif()
+
+if(JITSERVER_SUPPORT)
+	jvm_add_exports(JITServer_CreateServer)
```
The first argument should be the library name: `jvm_add_exports(jvm JITServer_CreateServer)`.

ashu-mehra replied:
Done

keithc-ca commented on runtime/vm/threadhelp.cpp:
```diff
@@ -298,37 +299,82 @@ getMonitorForWait(J9VMThread* vmThread, j9object_t object)
+        if (failedToSetAttr(omrthread_attr_set_schedpolicy(attr, J9THREAD_SCHEDPOLICY_OTHER))) {
+                rc = J9THREAD_ERR_INVALID_CREATE_ATTR;
+                goto _end;
+        }
```
Inconsistent indentation throughout this function.

keithc-ca commented on runtime/vm/threadhelp.cpp:
```diff
@@ -298,37 +299,82 @@ getMonitorForWait(J9VMThread* vmThread, j9object_t object)
+	if (J9THREAD_SUCCESS != rc) {
 		goto destroy_attr;
-	}
+	};
```
Please remove this added semicolon.

ashu-mehra replied:
Done

ashu-mehra commented:
I think I have addressed all the comments now. @keithc-ca can you please review it again.

Just for the record, there are two more PRs related to this https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/350 and https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/239 that are required for jitserver launcher to be packaged in the OpenJ9 builds at adoptopenjdk.

ashu-mehra commented:
I have tested these changes locally on linux_x86-64_compressedrefs with and without cmake on Java 8 and Java 11.

ashu-mehra commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1524,88 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+	rc = (*vm)->AttachCurrentThread(vm, (void **)&env, NULL);
+	if (JNI_OK == rc) {
+		J9JITConfig *jitConfig = ((J9JavaVM *)vm)->jitConfig;
+		rc = jitConfig->startJITServer(jitConfig);
```
`startJITServer` returns 0 on  success, -1 on error. I think its easier to manage if JIT return/error codes are kept separate from the JITServer interface return/error codes.

DanHeidinga replied:
It means we potentially lose information here - there is no path to find out why the startJITServer call failed

ashu-mehra replied:
I agree with that, but I think we should be adding tracing for this instead of trying to cover each and every failure that can happen.
Moreover, the Listener thread is not currently passing any error information down the stack.

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,354 @@
+		fprintf(stderr, "JITServer_CreateServer failed, error code=%d\n", rc);
+		goto _end;
+	}
+
```
Added this statement at the start of the launcher.

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,354 @@
+ 	/* pathBuffer can be either <jdk_home> or <jdk_home>/jre or <jre_home> */
+	char *lastLevel = strrchr(pathBuffer, DIR_SEPARATOR_CHAR);
+
+	if ((NULL != lastLevel) && !strcmp(lastLevel + 1, "jre")) {
```
Done

ashu-mehra replied:
btw I just realised it should be `(strcmp(lastLevel + 1, "jre") == 0)`

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,354 @@
+static int32_t 
+append(char **buffer, uint32_t bufLen, char *str)
+{
+	if (strlen(*buffer) + strlen(str) >= bufLen) {
```
Done

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,354 @@
+#endif
+
+	/* Consume any trailing '/' */
+	if (path[pathLen - 1] == DIR_SEPARATOR_CHAR) {
```
Done

ashu-mehra commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,354 @@
+
+	/* try "compressedrefs" first */
+	rc = append(&pathBuffer, pathBufLen, "/" OPENJ9_CR_JVM_DIR);
+        if (0 != rc) {
```
Fixed

ashu-mehra commented:
@DanHeidinga I have made the suggested changes.

ashu-mehra commented:
@DanHeidinga any more comments for this PR?

ashu-mehra commented on buildspecs/aix_ppc-64.spec:
```diff
@@ -128,6 +128,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<flag id="build_j2se" value="true"/>
 		<flag id="build_java8" value="true"/>
 		<flag id="build_java9" value="false"/>
+		<flag id="build_jitserver" value="false"/>
```
I added this flag in each spec in case we want to enable it later on other platforms.
Currently the openjdk build scripts rely on `<flag id="build_jitserver"` to be present to toggle it on when `--enable jitserver` is specified. See the changes in my PR https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/350.
In absence of these flags, I think it won't be very straight forward that we need to add `build_jitserver` flag in the spec file when enabling it for any new platform.

DanHeidinga replied:
This is the standard pattern we use for flags - they are, with some historical cruft aside, only added to the spec when when they are enabled.

ashu-mehra replied:
Are you suggesting to remove it from all specs because it not enabled by default in any. It gets set only if `--enable-jitserver` option is specified to configure step when building openjdk. Removing it from all specs would break ibmruntimes/openj9-openjdk-jdk8#350.

pshipton replied:
Looking at some of the other features, it seems you can add the line to the core.feature file and not modify every spec file.

ashu-mehra replied:
yes, this would work.

ashu-mehra commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1524,88 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+	}
+	server->startJITServer = startJITServer;
+	server->waitForJITServerTermination = waitForJITServerTermination;
+	rc = JNI_CreateJavaVM_impl(&server->jvm, (void **)&env, vm_args, TRUE);
```
Is there existing check to prevent multiple JVMs from being created in the same process?

As for the tests, I think [`vm_lifecycle`](https://github.com/eclipse/openj9/tree/master/runtime/tests/vm_lifecycle) tests seem to be appropriate place for doing these checks, however, currently they are disabled https://github.com/eclipse/openj9/blob/ecb16ff5c4fc581484f6c9acaa5e0891082c4423/test/functional/NativeTest/playlist.xml#L531.

I guess this PR can be merged without these checks for now. If that is ok, then I will address the checks in a separate issue, as it also needs the `vm_lifecycle` tests to be enabled again.

mpirvu replied:
Personally I don't think we need to add any check. It's the user responsability to not misuse an API.

pshipton replied:
The check occurs in the public JNI_CreateJavaVM() API in jvm.c using a static variable, the JNI_CreateJavaVM() in this file is internal.
```
	if (BFUjavaVM != NULL) {
		result = JNI_ERR;
		goto exit;
	}
```

pshipton replied:
Sorry, got confused about which file I was looking at. Anyway, JITServer_CreateServer() should add the above check.

ashu-mehra commented:
@DanHeidinga this seems to be waiting for couple of things that you raised for which we don't seem to have an immediate resolution. If those are not critical, can we go ahead with this PR and later on discuss how to address those points.

ashu-mehra commented:
@mpirvu @DanHeidinga is this still in 0.18 release plan?

mpirvu commented:
> @mpirvu @DanHeidinga is this still in 0.18 release plan?

Yes, the launcher is still in the 0.18.0 release plan

pshipton commented on buildspecs/linux_arm.spec:
```diff
@@ -136,6 +136,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<flag id="build_dropToHursley" value="true"/>
 		<flag id="build_j2se" value="true"/>
 		<flag id="build_java8" value="true"/>
+		<flag id="build_jitserver" value="true"/>
```
I don't expect build_jitserver should be set true for this platform.

ashu-mehra replied:
Right, fixed it.

pshipton commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -3757,6 +3757,10 @@ typedef struct J9JITConfig {
 	U_8* (*codeCacheWarmAlloc)(void *codeCache);
 	U_8* (*codeCacheColdAlloc)(void *codeCache);
+#if defined(JITSERVER_SUPPORT)
+	int32_t ( *startJITServer )(struct J9JITConfig *jitConfig);
```
Inconsistent spacing here and in the next change.

pshipton commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1524,88 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+
+#if defined(JITSERVER_SUPPORT)
+int32_t JNICALL 
+JITServer_CreateServer(JITServer **jitServer, void *vm_args)
```
New functions should have function comments.

pshipton commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,359 @@
+static int32_t getJvmPath(char *pathBuffer, int32_t pathBufLen);
+static char * getJvmLibPath(void);
+
+static const char *jvmOptions[] = {	"-Xms8m",
```
Hard coding options without a way to override them seems like a mistake. I'd add support for -J to pass options through.

ashu-mehra replied:
Currently the jitserver launcher accepts the same arguments that JVM does. I just add the set of `jvmOptions` to the user specified arguments and pass it to JVM. I think, to allow user to override `jvmOptions`, I can just add the user arguments at the end and JVM would pick the the one which comes last.

pshipton replied:
Sounds good.

pshipton commented on runtime/j9vm/jvm.c:
```diff
@@ -1514,6 +1524,88 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+}
+#endif /* JITSERVER_SUPPORT */
+
+static jint JNI_CreateJavaVM_impl(JavaVM **pvm, void **penv, void *vm_args, BOOLEAN isJITServer) {
```
Pls fix the function declaration to be on multiple lines.

pshipton commented:
Running some preliminary PR testing without JITSERVER_SUPPORT defined. It will keep running even if new changes are pushed. 

jenkins test sanity plinux,aix,osx,win,win32

pshipton commented:
jenkins test sanity plinux,aix,osx,win,win32 jdk8

pshipton commented:
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/2330/

pshipton commented:
The tests all failed, I'm thinking this needs a rebase.
Pls push a rebase separately from other changes, or make the changes in separate commits so it's easy to see what has changed. A squash can occur before merge.

pshipton commented:
Opened https://github.com/AdoptOpenJDK/openjdk-build/issues/1453 for the test issue.

ashu-mehra commented:
@pshipton Added a new commit with the suggested changes. Also rebased it. Please review.

pshipton commented on runtime/j9vm/jvm.c:
```diff
@@ -1558,6 +1556,13 @@ JITServer_CreateServer(JITServer **jitServer, void *vm_args)
 }
 
+/**
+ * Starts an instance of JITServer
```
Please end sentences with a period, in all the new comments.

pshipton commented on runtime/oti/jitserver_api.h:
```diff
@@ -37,7 +37,16 @@ typedef struct JITServer {
+ *
+ * @returns zero on success, else negative error code
+ */
+int32_t JNICALL JITServer_CreateServer(JITServer **jitServer, void *serverArgs);
```
This declaration should be on multiple lines.

pshipton commented on buildspecs/core.feature:
```diff
@@ -44,6 +44,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<project id="runtime"/>
 	</source>
 	<flags>
+		<flag id="build_jitserver" value="false"/>
```
The copyright check is failing on this file.

pshipton commented:
jenkins test sanity xlinuxjit jdk11

pshipton commented:
Not sure why the eca check is failing. This can be squashed.

pshipton commented:
There is a crash in the xlinuxjit build https://ci.eclipse.org/openj9/job/Build_JDK11_x86-64_linux_jit_Personal/20/console

ashu-mehra commented:
I did a xlinux build locally with jitserver enabled and it compiled without any issues. The crash at https://ci.eclipse.org/openj9/job/Build_JDK11_x86-64_linux_jit_Personal/20/console seems to be intermittent.

mpirvu commented:
jenkins test sanity xlinuxjit jdk8

mpirvu commented:
I realized that PR testing with this commit may not be very helpful, because the option that is used to launch the server, `-XX:StartAsJITServer` no longer exists.

ashu-mehra commented:
https://ci.eclipse.org/openj9/job/Build_JDK8_x86-64_linux_jit_Personal/27/console shows another crash with `java -version`. These changes should ideally not affect JVM's normal execution.

ashu-mehra commented:
Did couple of runs locally but I don't see the crash.

ashu-mehra commented:
In my personal build [1] for `x86-64_linux_jit` I see crash in jdk 11 compilation:
```
[2019-12-27T00:10:19.427Z] Unhandled exception
[2019-12-27T00:10:19.427Z] Type=Segmentation error vmState=0x00000000
[2019-12-27T00:10:19.427Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
[2019-12-27T00:10:19.427Z] Handler1=00002B91BF556090 Handler2=00002B91BFC8A090 InaccessibleAddress=0000000000000040
[2019-12-27T00:10:19.427Z] RDI=0000000000000000 RSI=0000000000000030 RAX=0000000000000030 RBX=00002B91CDC799A0
[2019-12-27T00:10:19.427Z] RCX=00002B91C00717A8 RDX=00002B91C00717E8 R8=00002B91C00717E8 R9=0000000000000000
[2019-12-27T00:10:19.427Z] R10=0000000000000000 R11=0000000000000000 R12=0000000000000000 R13=00002B91C00D1950
[2019-12-27T00:10:19.427Z] R14=0000000001FF6200 R15=0000000000000000
[2019-12-27T00:10:19.427Z] RIP=00002B91CADDFBDE GS=0000 FS=0000 RSP=00002B91DEF67780
[2019-12-27T00:10:19.427Z] EFlags=0000000000010202 CS=0033 RBP=0000000000000030 ERR=0000000000000004
[2019-12-27T00:10:19.427Z] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000040
[2019-12-27T00:10:19.427Z] xmm0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm1 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm4 4040404040404040 (f: 1077952512.000000, d: 3.250196e+01)
[2019-12-27T00:10:19.427Z] xmm5 5b5b5b5b5b5b5b5b (f: 1532713856.000000, d: 1.213625e+132)
[2019-12-27T00:10:19.427Z] xmm6 2020202020202020 (f: 538976256.000000, d: 6.013470e-154)
[2019-12-27T00:10:19.427Z] xmm7 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm8 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:19.427Z] xmm9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:19.427Z] xmm10 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:19.427Z] xmm11 000000004ad70a3c (f: 1255606784.000000, d: 6.203522e-315)
[2019-12-27T00:10:19.427Z] xmm12 000000004789a025 (f: 1200201728.000000, d: 5.929785e-315)
[2019-12-27T00:10:19.427Z] xmm13 0000000048ac0833 (f: 1219233792.000000, d: 6.023816e-315)
[2019-12-27T00:10:19.427Z] xmm14 0000000049650df0 (f: 1231359488.000000, d: 6.083724e-315)
[2019-12-27T00:10:19.427Z] xmm15 0000000047b73e60 (f: 1203191424.000000, d: 5.944555e-315)
[2019-12-27T00:10:19.427Z] Module=/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/lib/compressedrefs/libj9jit29.so
[2019-12-27T00:10:19.427Z] Module_base_address=00002B91CAC53000
[2019-12-27T00:10:19.427Z] Target=2_90_20191226_20 (Linux 2.6.32-754.24.3.el6.x86_64)
[2019-12-27T00:10:19.427Z] CPU=amd64 (8 logical CPUs) (0x1f409e000 RAM)
[2019-12-27T00:10:19.427Z] ----------- Stack Backtrace -----------
[2019-12-27T00:10:19.427Z] (0x00002B91CADDFBDE [libj9jit29.so+0x18cbde])
[2019-12-27T00:10:19.427Z] (0x00002B91CAF5FD94 [libj9jit29.so+0x30cd94])
[2019-12-27T00:10:19.427Z] (0x00002B91CADD6D43 [libj9jit29.so+0x183d43])
[2019-12-27T00:10:19.427Z] (0x00002B91CADD7192 [libj9jit29.so+0x184192])
[2019-12-27T00:10:19.427Z] (0x00002B91CADD723A [libj9jit29.so+0x18423a])
[2019-12-27T00:10:19.427Z] (0x00002B91BFC8ADF3 [libj9prt29.so+0x20df3])
[2019-12-27T00:10:19.427Z] (0x00002B91CADD7694 [libj9jit29.so+0x184694])
[2019-12-27T00:10:19.427Z] (0x00002B91BF8571A6 [libj9thr29.so+0xe1a6])
[2019-12-27T00:10:19.427Z] (0x00002B91BE00DAA1 [libpthread.so.0+0x7aa1])
[2019-12-27T00:10:19.427Z] clone+0x6d (0x00002B91BE721C4D [libc.so.6+0xe8c4d])
[2019-12-27T00:10:19.427Z] ---------------------------------------
[2019-12-27T00:10:19.427Z] JVMDUMP039I Processing dump event "gpf", detail "" at 2019/12/26 16:10:18 - please wait.
[2019-12-27T00:10:19.427Z] JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/core.20191226.161018.30853.0001.dmp' in response to an event
[2019-12-27T00:10:21.081Z] Unhandled exception
[2019-12-27T00:10:21.081Z] Type=Segmentation error vmState=0x00000000
[2019-12-27T00:10:21.081Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
[2019-12-27T00:10:21.081Z] Handler1=00002B91BF556090 Handler2=00002B91BFC8A090 InaccessibleAddress=0000000000000040
[2019-12-27T00:10:21.081Z] RDI=0000000000000000 RSI=0000000000000030 RAX=0000000000000030 RBX=00002B91CDC794B0
[2019-12-27T00:10:21.081Z] RCX=00002B91C00717A8 RDX=00002B91C00717E8 R8=00002B91C00717E8 R9=0000000000000000
[2019-12-27T00:10:21.081Z] R10=0000000000000000 R11=0000000000000202 R12=0000000000000000 R13=00002B91C00D1950
[2019-12-27T00:10:21.081Z] R14=0000000001FF2600 R15=0000000000000000
[2019-12-27T00:10:21.081Z] RIP=00002B91CADDFBDE GS=0000 FS=0000 RSP=00002B91DDE66780
[2019-12-27T00:10:21.081Z] EFlags=0000000000010202 CS=0033 RBP=0000000000000030 ERR=0000000000000004
[2019-12-27T00:10:21.081Z] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000040
[2019-12-27T00:10:21.081Z] xmm0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.081Z] xmm1 0a0a0a0a0a0a0a0a (f: 168430096.000000, d: 2.646194e-260)
[2019-12-27T00:10:21.081Z] xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.081Z] xmm3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.081Z] xmm4 4040404040404040 (f: 1077952512.000000, d: 3.250196e+01)
[2019-12-27T00:10:21.081Z] xmm5 5b5b5b5b5b5b5b5b (f: 1532713856.000000, d: 1.213625e+132)
[2019-12-27T00:10:21.081Z] xmm6 2020202020202020 (f: 538976256.000000, d: 6.013470e-154)
[2019-12-27T00:10:21.081Z] xmm7 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.081Z] xmm8 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:21.081Z] xmm9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.081Z] xmm10 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:21.081Z] xmm11 000000004ad70a3c (f: 1255606784.000000, d: 6.203522e-315)
[2019-12-27T00:10:21.081Z] xmm12 000000004789a025 (f: 1200201728.000000, d: 5.929785e-315)
[2019-12-27T00:10:21.081Z] xmm13 0000000048ac0833 (f: 1219233792.000000, d: 6.023816e-315)
[2019-12-27T00:10:21.081Z] xmm14 0000000049650df0 (f: 1231359488.000000, d: 6.083724e-315)
[2019-12-27T00:10:21.081Z] xmm15 0000000047b73e60 (f: 1203191424.000000, d: 5.944555e-315)
[2019-12-27T00:10:21.081Z] Module=/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/lib/compressedrefs/libj9jit29.so
[2019-12-27T00:10:21.081Z] Module_base_address=00002B91CAC53000
[2019-12-27T00:10:21.081Z] Target=2_90_20191226_20 (Linux 2.6.32-754.24.3.el6.x86_64)
[2019-12-27T00:10:21.081Z] CPU=amd64 (8 logical CPUs) (0x1f409e000 RAM)
[2019-12-27T00:10:21.081Z] ----------- Stack Backtrace -----------
[2019-12-27T00:10:21.081Z] (0x00002B91CADDFBDE [libj9jit29.so+0x18cbde])
[2019-12-27T00:10:21.081Z] (0x00002B91CAF5FD94 [libj9jit29.so+0x30cd94])
[2019-12-27T00:10:21.081Z] (0x00002B91CADD6D43 [libj9jit29.so+0x183d43])
[2019-12-27T00:10:21.081Z] (0x00002B91CADD7192 [libj9jit29.so+0x184192])
[2019-12-27T00:10:21.081Z] (0x00002B91CADD723A [libj9jit29.so+0x18423a])
[2019-12-27T00:10:21.081Z] (0x00002B91BFC8ADF3 [libj9prt29.so+0x20df3])
[2019-12-27T00:10:21.081Z] (0x00002B91CADD7694 [libj9jit29.so+0x184694])
[2019-12-27T00:10:21.081Z] (0x00002B91BF8571A6 [libj9thr29.so+0xe1a6])
[2019-12-27T00:10:21.081Z] (0x00002B91BE00DAA1 [libpthread.so.0+0x7aa1])
[2019-12-27T00:10:21.081Z] clone+0x6d (0x00002B91BE721C4D [libc.so.6+0xe8c4d])
[2019-12-27T00:10:21.081Z] ---------------------------------------
[2019-12-27T00:10:21.082Z] JVMDUMP039I Processing dump event "gpf", detail "" at 2019/12/26 16:10:20 - please wait.
[2019-12-27T00:10:21.082Z] Unhandled exception
[2019-12-27T00:10:21.082Z] Type=Segmentation error vmState=0x00000000
[2019-12-27T00:10:21.082Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
[2019-12-27T00:10:21.082Z] Handler1=00002B91BF556090 Handler2=00002B91BFC8A090 InaccessibleAddress=0000000000000040
[2019-12-27T00:10:21.082Z] RDI=0000000000000000 RSI=0000000000000030 RAX=0000000000000030 RBX=00002B91CDC79E90
[2019-12-27T00:10:21.082Z] RCX=00002B91C00717A8 RDX=00002B91C00717E8 R8=00002B91C00717E8 R9=0000000000000000
[2019-12-27T00:10:21.082Z] R10=0000000000000000 R11=0000000000000202 R12=0000000000000000 R13=00002B91C00D1950
[2019-12-27T00:10:21.082Z] R14=0000000001FF9E00 R15=0000000000000000
[2019-12-27T00:10:21.082Z] RIP=00002B91CADDFBDE GS=0000 FS=0000 RSP=00002B91E80FD780
[2019-12-27T00:10:21.082Z] EFlags=0000000000010202 CS=0033 RBP=0000000000000030 ERR=0000000000000004
[2019-12-27T00:10:21.082Z] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000040
[2019-12-27T00:10:21.082Z] xmm0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.082Z] xmm1 0a0a0a0a0a0a0a0a (f: 168430096.000000, d: 2.646194e-260)
[2019-12-27T00:10:21.082Z] xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.082Z] xmm3 ffffff0000000000 (f: 0.000000, d: -nan)
[2019-12-27T00:10:21.082Z] xmm4 4040404040404040 (f: 1077952512.000000, d: 3.250196e+01)
[2019-12-27T00:10:21.082Z] xmm5 5b5b5b5b5b5b5b5b (f: 1532713856.000000, d: 1.213625e+132)
[2019-12-27T00:10:21.082Z] xmm6 2020202020202020 (f: 538976256.000000, d: 6.013470e-154)
[2019-12-27T00:10:21.082Z] xmm7 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.082Z] xmm8 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:21.082Z] xmm9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:21.082Z] xmm10 00ff000000ff0000 (f: 16711680.000000, d: 7.063274e-304)
[2019-12-27T00:10:21.082Z] xmm11 000000004ad70a3c (f: 1255606784.000000, d: 6.203522e-315)
[2019-12-27T00:10:21.082Z] xmm12 000000004789a025 (f: 1200201728.000000, d: 5.929785e-315)
[2019-12-27T00:10:21.082Z] xmm13 0000000048ac0833 (f: 1219233792.000000, d: 6.023816e-315)
[2019-12-27T00:10:21.082Z] xmm14 0000000049650df0 (f: 1231359488.000000, d: 6.083724e-315)
[2019-12-27T00:10:21.082Z] xmm15 0000000047b73e60 (f: 1203191424.000000, d: 5.944555e-315)
[2019-12-27T00:10:21.082Z] Module=/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/lib/compressedrefs/libj9jit29.so
[2019-12-27T00:10:21.082Z] Module_base_address=00002B91CAC53000
[2019-12-27T00:10:21.082Z] Target=2_90_20191226_20 (Linux 2.6.32-754.24.3.el6.x86_64)
[2019-12-27T00:10:21.082Z] CPU=amd64 (8 logical CPUs) (0x1f409e000 RAM)
[2019-12-27T00:10:21.082Z] ----------- Stack Backtrace -----------
[2019-12-27T00:10:21.082Z] (0x00002B91CADDFBDE [libj9jit29.so+0x18cbde])
[2019-12-27T00:10:21.082Z] (0x00002B91CAF5FD94 [libj9jit29.so+0x30cd94])
[2019-12-27T00:10:21.082Z] (0x00002B91CADD6D43 [libj9jit29.so+0x183d43])
[2019-12-27T00:10:21.082Z] (0x00002B91CADD7192 [libj9jit29.so+0x184192])
[2019-12-27T00:10:21.082Z] (0x00002B91CADD723A [libj9jit29.so+0x18423a])
[2019-12-27T00:10:21.082Z] (0x00002B91BFC8ADF3 [libj9prt29.so+0x20df3])
[2019-12-27T00:10:21.082Z] (0x00002B91CADD7694 [libj9jit29.so+0x184694])
[2019-12-27T00:10:21.082Z] (0x00002B91BF8571A6 [libj9thr29.so+0xe1a6])
[2019-12-27T00:10:21.082Z] (0x00002B91BE00DAA1 [libpthread.so.0+0x7aa1])
[2019-12-27T00:10:21.082Z] clone+0x6d (0x00002B91BE721C4D [libc.so.6+0xe8c4d])
[2019-12-27T00:10:21.082Z] ---------------------------------------
[2019-12-27T00:10:21.082Z] JVMDUMP039I Processing dump event "gpf", detail "" at 2019/12/26 16:10:20 - please wait.
[2019-12-27T00:10:21.082Z] JVMDUMP010I System dump written to /home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/core.20191226.161018.30853.0001.dmp
[2019-12-27T00:10:21.082Z] JVMDUMP032I JVM requested Java dump using '/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/javacore.20191226.161018.30853.0004.txt' in response to an event
[2019-12-27T00:10:21.082Z] 
[2019-12-27T00:10:21.082Z] 
[2019-12-27T00:10:21.082Z] *** Invalid JIT return address 00000000FFB65CE0 in 00002B91DEF60B80
[2019-12-27T00:10:21.082Z] 
[2019-12-27T00:10:21.082Z] JVMDUMP010I Java dump written to /home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/javacore.20191226.161018.30853.0004.txt
[2019-12-27T00:10:21.082Z] JVMDUMP032I JVM requested Snap dump using '/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/Snap.20191226.161018.30853.0005.trc' in response to an event
[2019-12-27T00:10:21.885Z] JVMDUMP010I Snap dump written to /home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/Snap.20191226.161018.30853.0005.trc
[2019-12-27T00:10:21.885Z] JVMDUMP007I JVM Requesting JIT dump using '/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/jitdump.20191226.161018.30853.0006.dmp'
[2019-12-27T00:10:21.885Z] JVMDUMP010I JIT dump written to /home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/make/jitdump.20191226.161018.30853.0006.dmp
[2019-12-27T00:10:21.885Z] JVMDUMP013I Processed dump event "gpf", detail "".
[2019-12-27T00:10:21.885Z] gmake[3]: *** [/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/_packages_attribute.done] Error 1
[2019-12-27T00:10:21.885Z] ExplodedImageOptimize.gmk:40: recipe for target '/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/_packages_attribute.done' failed
[2019-12-27T00:10:21.885Z] gmake[2]: *** [exploded-image-optimize] Error 1
[2019-12-27T00:10:21.885Z] make/Main.gmk:351: recipe for target 'exploded-image-optimize' failed
[2019-12-27T00:10:21.885Z] gmake[2]: *** Waiting for unfinished jobs....
[2019-12-27T00:10:48.092Z] 
[2019-12-27T00:10:48.092Z] ERROR: Build failed for target 'all' in configuration 'linux-x86_64-normal-server-release' (exit code 2) 
[2019-12-27T00:10:48.092Z] Stopping sjavac server
[2019-12-27T00:10:48.092Z] 
[2019-12-27T00:10:48.092Z] === Output from failing command(s) repeated here ===
[2019-12-27T00:10:48.092Z] * For target jdk__packages_attribute.done:
[2019-12-27T00:10:48.092Z] Unhandled exception
[2019-12-27T00:10:48.092Z] Type=Segmentation error vmState=0x00000000
[2019-12-27T00:10:48.092Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
[2019-12-27T00:10:48.092Z] Handler1=00002B91BF556090 Handler2=00002B91BFC8A090 InaccessibleAddress=0000000000000040
[2019-12-27T00:10:48.092Z] RDI=0000000000000000 RSI=0000000000000030 RAX=0000000000000030 RBX=00002B91CDC799A0
[2019-12-27T00:10:48.092Z] RCX=00002B91C00717A8 RDX=00002B91C00717E8 R8=00002B91C00717E8 R9=0000000000000000
[2019-12-27T00:10:48.092Z] R10=0000000000000000 R11=0000000000000000 R12=0000000000000000 R13=00002B91C00D1950
[2019-12-27T00:10:48.092Z] R14=0000000001FF6200 R15=0000000000000000
[2019-12-27T00:10:48.092Z] RIP=00002B91CADDFBDE GS=0000 FS=0000 RSP=00002B91DEF67780
[2019-12-27T00:10:48.092Z] EFlags=0000000000010202 CS=0033 RBP=0000000000000030 ERR=0000000000000004
[2019-12-27T00:10:48.092Z] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000040
[2019-12-27T00:10:48.092Z] xmm0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:48.092Z] xmm1 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:48.092Z] xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:48.092Z] xmm3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2019-12-27T00:10:48.092Z]    ... (rest of output omitted)
[2019-12-27T00:10:48.092Z] 
[2019-12-27T00:10:48.092Z] * All command lines available in /home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/make-support/failure-logs.
[2019-12-27T00:10:48.092Z] === End of repeated output ===
```

From the javacore, which is available in the diagnostic artifact [2], I see there are 63 JIT Compilation threads which implies that the JVM was running in server mode. Also the executable of the core dump is `java` launcher (`/home/jenkins/workspace/Build_JDK11_x86-64_linux_jit_Personal/build/linux-x86_64-normal-server-release/jdk/bin/java`)
However, with this PR it should not be possible to start JVM in server without using `jitserver` launcher.

[1] https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK11_x86-64_linux_jit_Personal/20/console
[2] https://na.artifactory.swg-devops.com/artifactory/sys-rt-generic-local/hyc-runtimes-jenkins.swg-devops.com/Build_JDK11_x86-64_linux_jit_Personal/20/Build_JDK11_x86-64_linux_jit_Personal-20-20191226-185437-diagnostics.tar.gz

ashu-mehra commented:
I realised the crash in PR builds is because the openjdk makefile changes [1] to enable JITSERVER_SUPPORT macro in VM have not been merged yet. This causes JITServer support to be enabled in JIT but disabled in VM, causing these weird crashes.
I verified this by running two personal builds. The build [2] that used openj9 branch of `openj9-openjdk-jdk11` crashed in the same manner as PR builds. The other build [3] used my branch which has [1] and it passed without any issues.

I think once the PRs https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/239 and https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/350 are merged, these crashes would go away.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/239
[2] https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/4392/
[3] https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/4397/

pshipton commented:
jenkins test sanity xlinux,plinux,aix,osx,win,win32 jdk8

pshipton commented:
jenkins test sanity xlinuxjit jdk8 depends ibmruntimes/openj9-openjdk-jdk8#350

pshipton commented:
jenkins test sanity xlinuxjit jdk11 depends ibmruntimes/openj9-openjdk-jdk11#239

pshipton commented:
jenkins test sanity xlinuxcm jdk8

ashu-mehra commented:
One test `cmdLineTester_getPid_0` failed which is expected as the test is not handling the technology preview message displayed by the client JVM connecting to a JITServer.

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,364 @@
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !(defined(LINUX) && !defined(J9ZTPF) && !defined(J9ARM) && !defined(J9AARCH64))
```
I think this would be easier to read if the negation was pushed down:
```
#if !defined(LINUX) || defined(J9ZTPF) || defined(J9ARM) || defined(J9AARCH64)
```

ashu-mehra replied:
Done

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,364 @@
+}
+
+static int32_t
+isDir(const char *path) {
```
formatting: brace on a separate line; also on lines 122 & 134 below

ashu-mehra replied:
Done

keithc-ca commented on runtime/jitserver_launcher/jitserver.c:
```diff
@@ -0,0 +1,364 @@
+	return jvmLibPath;
+}
+
+int main(int argc, char *argv[])
```
formatting: return type on a separate line

ashu-mehra replied:
Done

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -597,6 +597,13 @@ areValueTypesEnabled(J9JavaVM *vm)
 
+#if defined(JITSERVER_SUPPORT)
+BOOLEAN
+isJITServerEnabled(J9JavaVM *vm) {
```
formatting: brace on a separate line

ashu-mehra replied:
Done

keithc-ca commented:
Please fix the copyright dates and address the merge conflict.

ashu-mehra commented:
Fixed the copyright dates.

keithc-ca commented:
Once ibmruntimes/openj9-openjdk-jdk11#239 has been finalized, we'll want similar changes for jdk13, jdk14 and jdknext before this should be merged otherwise the problem mentioned [above](https://github.com/eclipse/openj9/pull/7940#issuecomment-569185608) will occur for those versions.

ashu-mehra commented:
> we'll want similar changes for jdk13, jdk14 and jdknext 

@keithc-ca I will create PRs for jdk13 and 14, but I can't find any `jdknext` repo at https://github.com/ibmruntimes?
utf8=%E2%9C%93&q=openj9-openjdk-jdk&type=&language=

ashu-mehra commented:
I guess we don't need this for jdk12 as there are no Adopt builds and it is already out of support, right?

pshipton commented:
Correct about jdk12, it is obsolete now.
jdk "next" is https://github.com/ibmruntimes/openj9-openjdk-jdk

ashu-mehra commented:
Raised PRs to include jitserver launcher in the builds for all extension repo: 
https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/157 
https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/350
https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/239
https://github.com/ibmruntimes/openj9-openjdk-jdk13/pull/33
https://github.com/ibmruntimes/openj9-openjdk-jdk14/pull/1

keithc-ca commented:
Jenkins test sanity xlinuxjit jdk8 depends ibmruntimes/openj9-openjdk-jdk8#350

keithc-ca commented:
Jenkins test sanity xlinuxjit jdk11 depends ibmruntimes/openj9-openjdk-jdk11#239

keithc-ca commented:
Jenkins test sanity xlinuxjit jdk13 depends ibmruntimes/openj9-openjdk-jdk13#33

keithc-ca commented:
The changes for jdk14 and jdknext are identical to those for jdk13: they shouldn't need separate testing.

keithc-ca commented:
The request for a jdk13 build was rejected: see #8209.

ashu-mehra commented:
Both jdk8 and jdk11 sanity tests are crashing during compilation of test material:
https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_jit_Personal/12/consoleText
https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/12/consoleText
```
[2020-01-06T17:50:40.562Z] dist:
[2020-01-06T17:50:40.562Z]       [jar] Building jar: /home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/jvmtest/functional/cmdLineTests/shareClassTests/URLHelperTests/URLHelperTests.jar
[2020-01-06T17:50:41.684Z] Unhandled exception
[2020-01-06T17:50:41.684Z] Type=Segmentation error vmState=0x00020011
[2020-01-06T17:50:41.684Z] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
[2020-01-06T17:50:41.684Z] Handler1=00007F4335FBC010 Handler2=00007F43358A7090 InaccessibleAddress=0000000000000008
[2020-01-06T17:50:41.684Z] RDI=0000000000000000 RSI=00000000025D5300 RAX=0000000000000000 RBX=0000000000000000
[2020-01-06T17:50:41.684Z] RCX=0000000000000007 RDX=0000000000000002 R8=0000000000000001 R9=0000000000000000
[2020-01-06T17:50:41.684Z] R10=0000000000000000 R11=0000000000000246 R12=00000000025D5300 R13=00007F43300AB268
[2020-01-06T17:50:41.684Z] R14=00000000025D5300 R15=00007F43283CA010
[2020-01-06T17:50:41.684Z] RIP=00007F43294086EE GS=0000 FS=0000 RSP=00007F43370F2360
[2020-01-06T17:50:41.684Z] EFlags=0000000000010206 CS=0033 RBP=00000000025D5300 ERR=0000000000000004
[2020-01-06T17:50:41.684Z] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000008
[2020-01-06T17:50:41.684Z] xmm0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2020-01-06T17:50:41.684Z] xmm1 0000000000ff0000 (f: 16711680.000000, d: 8.256667e-317)
[2020-01-06T17:50:41.684Z] xmm2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2020-01-06T17:50:41.684Z] xmm3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
[2020-01-06T17:50:41.684Z] xmm4 3d69237861bbcc89 (f: 1639697536.000000, d: 7.144807e-13)
[2020-01-06T17:50:41.684Z] xmm5 3ff413ff30000000 (f: 805306368.000000, d: 1.254882e+00)
[2020-01-06T17:50:41.684Z] xmm6 3fe9801a88459b21 (f: 2286263040.000000, d: 7.968877e-01)
[2020-01-06T17:50:41.684Z] xmm7 3bbcc86800000000 (f: 0.000000, d: 6.095003e-21)
[2020-01-06T17:50:41.684Z] xmm8 402279d53a3bcdd6 (f: 976997824.000000, d: 9.237955e+00)
[2020-01-06T17:50:41.684Z] xmm9 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
[2020-01-06T17:50:41.684Z] xmm10 3eaa000000000000 (f: 0.000000, d: 7.748604e-07)
[2020-01-06T17:50:41.684Z] xmm11 3d6925b7281203d7 (f: 672269248.000000, d: 7.147300e-13)
[2020-01-06T17:50:41.684Z] xmm12 3ea4b8158eb88e0b (f: 2394459648.000000, d: 6.174767e-07)
[2020-01-06T17:50:41.684Z] xmm13 3b5296b7a871af60 (f: 2826022656.000000, d: 6.150498e-23)
[2020-01-06T17:50:41.684Z] xmm14 402205966f2b4d80 (f: 1865108864.000000, d: 9.010913e+00)
[2020-01-06T17:50:41.684Z] xmm15 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
[2020-01-06T17:50:41.684Z] Module=/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdkbinary/j2sdk-image/jre/lib/amd64/compressedrefs/libj9jit29.so
[2020-01-06T17:50:41.684Z] Module_base_address=00007F4329190000
[2020-01-06T17:50:41.684Z] Target=2_90_20200106_31 (Linux 2.6.32-754.11.1.el6.x86_64)
[2020-01-06T17:50:41.684Z] CPU=amd64 (4 logical CPUs) (0x1ec001000 RAM)
[2020-01-06T17:50:41.684Z] ----------- Stack Backtrace -----------
[2020-01-06T17:50:41.684Z] (0x00007F43294086EE [libj9jit29.so+0x2786ee])
[2020-01-06T17:50:41.684Z] (0x00007F4329408819 [libj9jit29.so+0x278819])
[2020-01-06T17:50:41.684Z] (0x00007F4329408E27 [libj9jit29.so+0x278e27])
[2020-01-06T17:50:41.684Z] (0x00007F432932989E [libj9jit29.so+0x19989e])
[2020-01-06T17:50:41.684Z] (0x00007F4335B0339E [libj9hookable29.so+0x139e])
[2020-01-06T17:50:41.684Z] (0x00007F4334BC8945 [libj9gc29.so+0x1dc945])
[2020-01-06T17:50:41.684Z] (0x00007F4334BC8B42 [libj9gc29.so+0x1dcb42])
[2020-01-06T17:50:41.684Z] (0x00007F4334B566C5 [libj9gc29.so+0x16a6c5])
[2020-01-06T17:50:41.684Z] (0x00007F4334B5687E [libj9gc29.so+0x16a87e])
[2020-01-06T17:50:41.684Z] (0x00007F4334B6164B [libj9gc29.so+0x17564b])
[2020-01-06T17:50:41.684Z] (0x00007F4334BAFEF7 [libj9gc29.so+0x1c3ef7])
[2020-01-06T17:50:41.684Z] (0x00007F4334B16D67 [libj9gc29.so+0x12ad67])
[2020-01-06T17:50:41.684Z] (0x00007F4334BB64E6 [libj9gc29.so+0x1ca4e6])
[2020-01-06T17:50:41.684Z] (0x00007F4334BB8537 [libj9gc29.so+0x1cc537])
[2020-01-06T17:50:41.684Z] (0x00007F4334BB8B48 [libj9gc29.so+0x1ccb48])
[2020-01-06T17:50:41.684Z] (0x00007F4334B5B8A9 [libj9gc29.so+0x16f8a9])
[2020-01-06T17:50:41.684Z] (0x00007F4334A3196F [libj9gc29.so+0x4596f])
[2020-01-06T17:50:41.684Z] (0x00007F4335F48EAF [libj9vm29.so+0x25eaf])
[2020-01-06T17:50:41.684Z] (0x00007F4335F356E0 [libj9vm29.so+0x126e0])
[2020-01-06T17:50:41.684Z] (0x00007F4335FF6E22 [libj9vm29.so+0xd3e22])
[2020-01-06T17:50:41.684Z] ---------------------------------------
[2020-01-06T17:50:41.685Z] JVMDUMP039I Processing dump event "gpf", detail "" at 2020/01/06 12:50:25 - please wait.
[2020-01-06T17:50:41.685Z] JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/core.20200106.125025.26116.0001.dmp' in response to an event
[2020-01-06T17:51:04.665Z] JVMPORT030W /proc/sys/kernel/core_pattern setting "|/usr/libexec/abrt-hook-ccpp %s %c %p %u %g %t e" specifies that the core dump is to be piped to an external program.  Attempting to rename either core or core.6348.
[2020-01-06T17:51:04.665Z] 
[2020-01-06T17:51:04.665Z] JVMDUMP010I System dump written to /home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/core.20200106.125025.26116.0001.dmp
[2020-01-06T17:51:04.665Z] JVMDUMP032I JVM requested Java dump using '/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/javacore.20200106.125025.26116.0002.txt' in response to an event
[2020-01-06T17:51:04.665Z] JVMDUMP010I Java dump written to /home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/javacore.20200106.125025.26116.0002.txt
[2020-01-06T17:51:04.665Z] JVMDUMP032I JVM requested Snap dump using '/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/Snap.20200106.125025.26116.0003.trc' in response to an event
[2020-01-06T17:51:04.665Z] JVMDUMP010I Snap dump written to /home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/Snap.20200106.125025.26116.0003.trc
[2020-01-06T17:51:04.665Z] JVMDUMP007I JVM Requesting JIT dump using '/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/jitdump.20200106.125025.26116.0004.dmp'
[2020-01-06T17:51:04.665Z] JVMDUMP010I JIT dump written to /home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/jitdump.20200106.125025.26116.0004.dmp
[2020-01-06T17:51:04.665Z] JVMDUMP013I Processed dump event "gpf", detail "".
[2020-01-06T17:51:04.665Z] 255
[2020-01-06T17:51:04.665Z] 20200106-12:50:48 - Attempting to move the TDUMPS to '/home/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/openjdk-tests/TKG/../TKG/test_output_157833192331/compile', using the log to identify the TDUMP name to be moved
```

This does not look related to changes in this PR. Using the [Java8 PR build](https://140-211-168-230-openstack.osuosl.org/artifactory/ci-eclipse-openj9/Build_JDK8_x86-64_linux_jit_Personal/31/OpenJ9-JDK8-x86-64_linux_jit-20200106-162055.tar.gz) I compiled the test material 3 times and it didn't crash.
I am also running a [personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/4425/) which has gone past the crashing point seen in PR builds.

ashu-mehra commented:
Top frames of the crash mentioned in previous comment correspond to:
```
TR_AddressSet::firstHigherAddressRangeIndex(uintptrj_t address)
TR_AddressSet::add(uintptrj_t address)
J9::PersistentInfo::addUnloadedClass(TR_OpaqueClassBlock *clazz, uintptrj_t startAddress, uint32_t size)
jitHookClassUnload()
J9HookDispatch()
MM_ClassLoaderManager::addDyingClassesToList
```

Crashing instruction is `libj9jit29.so+0x2786ee`:
```
00000000002786e0 <_ZN13TR_AddressSet28firstHigherAddressRangeIndexEm>:
  2786e0:       41 57                   push   %r15
  2786e2:       41 56                   push   %r14
  2786e4:       41 55                   push   %r13
  2786e6:       41 54                   push   %r12
  2786e8:       55                      push   %rbp
  2786e9:       53                      push   %rbx
  2786ea:       48 83 ec 08             sub    $0x8,%rsp
  2786ee:       44 8b 6f 08             mov    0x8(%rdi),%r13d
```

`rdi is 0.`

Corresponding code:
```
int32_t TR_AddressSet::firstHigherAddressRangeIndex(uintptrj_t address)
   {
   if (_numAddressRanges == 0)
      return _numAddressRanges;
   ...
```

`rdi` should be object of type `TR_AddressSet` corresponding to `J9::PersistentInfo::_unloadedClassAddresses` which should have been initialized in `J9::PersistentInfo::ensureUnloadedAddressSetsAreInitialized()`.

ashu-mehra commented:
It looks like while rebasing the PR to master I introduced a bug.

ashu-mehra commented:
@mpirvu can you please review this [commit](https://github.com/eclipse/openj9/pull/7940/commits/a670646980527dc049cbc01618e8d6869f0eea69) . It fixes a bug that got in while rebasing the PR.

ashu-mehra commented:
@keithc-ca can you please trigger the builds again?

keithc-ca commented:
Jenkins test sanity xlinuxjit jdk8 depends ibmruntimes/openj9-openjdk-jdk8#350

keithc-ca commented:
Jenkins test sanity xlinuxjit jdk11 depends ibmruntimes/openj9-openjdk-jdk11#239

ashu-mehra commented:
Java 11 test `cmdLineTester_getPid_0` failure is [expected](https://github.com/eclipse/openj9/pull/7940#issuecomment-569772242) and can be ignored.
All other tests have passed in https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_x86-64_linux_jit_Personal/13/
and https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_jit_Personal/13/

Squashing the commits as the pr builds have passed.

keithc-ca commented:
> Java 11 test `cmdLineTester_getPid_0` failure is [expected](https://github.com/eclipse/openj9/pull/7940#issuecomment-569772242) and can be ignored.
> All other tests have passed in ...

Can you also open an issue to address that test failure (and share a link to it here)?

DanHeidinga commented on runtime/j9vm/jvm.c:
```diff
@@ -1496,24 +1506,131 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+
+#if defined(JITSERVER_SUPPORT)
+int32_t JNICALL
+JITServer_CreateServer(JITServer **jitServer, void *serverArgs)
```
This should have a Doxygen style comment as well

ashu-mehra replied:
The comment is present in `oti/jitserver_api.h`.

DanHeidinga commented on buildspecs/core.feature:
```diff
@@ -44,6 +44,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<project id="runtime"/>
 	</source>
 	<flags>
+		<flag id="build_jitserver" value="false"/>
```
Please remove this line.  The flag defaults to false so this is redundant.

ashu-mehra replied:
This is required. openjdk-openj9 builds rely on turning `value="false"` to `value="true"` when `--enable-jitserver` is specified. As an example, see the change in jdk11 extensions repo here: https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/239/files.
Without this flag openjdk-openj9 builds won't be able to turn on jitserver support in OpenJ9 VM.

DanHeidinga replied:
This is an unfortunate design pattern but given there's precedent for it, I remove my objection

DanHeidinga commented on runtime/j9vm/jvm.c:
```diff
@@ -1496,24 +1506,131 @@ static jint initializeReflectionGlobals(JNIEnv * env, BOOLEAN includeAccessors)
+
+_end:
+	if ((JITSERVER_OK != rc) && (NULL != server)) {
+		free(server);
```
When does the `server = malloc(sizeof(JITServer));` get freed in the success case?  I don't see it in the `waitForJITServerTermination` call.  Is there is a missing `JITServer_DestroyServer` call?

DanHeidinga replied:
Also, I don't see where the JavaVM is destroyed...

ashu-mehra replied:
`JITServer_DestroyServer` is not there, as there is no way to call it for now. The main thread just keeps waiting for the JITServer listener thread which does not exit. To terminate JITServer, we need to kill the process.
This is not ideal, but it has been like this. As per my discussion with @mpirvu  we may add support to gracefully terminate the server in future.

DanHeidinga replied:
Please open an issue for adding the `JITServer_DestroyServer` api.  This PR can be merged once the issue is opened for this.

DanHeidinga commented on runtime/include/j9cfg.h.ftl:
```diff
@@ -70,6 +70,12 @@ extern "C" {
 
+#if defined(J9VM_BUILD_JITSERVER)
+#ifndef JITSERVER_SUPPORT
+#define JITSERVER_SUPPORT
```
This feels off to me after looking at it again.  Why are we defining a flag to define a different flag?

The usual way to do this in J9 is to define an `opt_` flag and then use the `J9VM_OPT_JITSERVER` flag everywhere.  I think it would be cleaner to have 1 consistent flag controlled by the `j9.flags` file then to have that flag set this one.

ashu-mehra replied:
Reason to do it this way is `JITSERVER_SUPPORT` is already being used by JIT to enable the related code. And if I remember correctly @keithc-ca wanted to use same macro for VM as well, and this was the easiest way to make it all work (in both cmake and uma build systems).

DanHeidinga replied:
Please open an issue to address this in the future.  There should be a single consistent option used throughout the code base to enable / disable jitserver

ashu-mehra commented:
> Can you also open an issue to address that test failure (and share a link to it here)?

Issue for failing test: https://github.com/eclipse/openj9/issues/8220

ashu-mehra commented:
If there are no more comments, then it should be ready to be merged once the [extension repo PRs](https://github.com/eclipse/openj9/pull/7940#issuecomment-570735120) are merged.

keithc-ca commented:
This and the extension repo PRs must be merged together; merging any subset creates an inconsistent state.

DanHeidinga commented:
I'm OK with this being merged once the two requested issues are opened.

ashu-mehra commented:
Issue to track `JITServer_DestroyServer` api: https://github.com/eclipse/openj9/issues/8228

ashu-mehra commented:
Issue for using single flag to enable JITServer: https://github.com/eclipse/openj9/issues/8229

mpirvu commented:
@keithc-ca I think that all reviewers are satisfied with the current state of this PR. Could you please merge this PR and the dependent extension repo PRs? 
Thanks!

keithc-ca commented:
> @keithc-ca I think that all reviewers are satisfied with the current state of this PR. 

@DanHeidinga could you please confirm by approving this?

DanHeidinga commented:
Approved.  Sorry, didn't see @keithc-ca's comment till now

pshipton commented:
@ashu-mehra pls cherry pick this commit and create a PR against the v0.18.0-release branch.
